pkg/reloader/reloader.go

```diff
@@ -0,0 +1,212 @@
+// Package reloader contains helpers to trigger reloads of Prometheus instances
+// on configuration changes and to substitude environment variables in config files.
+package reloader
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"hash"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"time"
+
+	"github.com/a8m/envsubst"
+
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	fsnotify "gopkg.in/fsnotify.v1"
+)
+
+// Reloader can watch config files and trigger reloads of a Prometheus server.
+// It optionally substitutes environment variables in the configuration.
+type Reloader struct {
+	logger           log.Logger
+	promURL          *url.URL
+	cfgFilename      string
+	cfgSubstFilename string
+	ruleDir          string
+	lastCfgHash      []byte
+	lastRulesHash    []byte
+}
+
+// New creates a new reloader that watches the given input file and copies
+// it to the output file path.
+// It optionally substitutes environment variables on copying.
+func New(logger log.Logger, promURL *url.URL, cfgFile, cfgEnvsubstFile, ruleDir string) *Reloader {
+	return &Reloader{
+		logger:           logger,
+		promURL:          promURL,
+		cfgFilename:      cfgFile,
+		cfgSubstFilename: cfgEnvsubstFile,
+	}
+}
+
+// Watch starts to watch the config file and processes it until the context
+// gets canceled.
+func (r *Reloader) Watch(ctx context.Context) error {
+	w, err := fsnotify.NewWatcher()
+	if err != nil {
+		return errors.Wrap(err, "create watcher")
+	}
+	defer w.Close()
+
+	cfgDir := filepath.Dir(r.cfgFilename)
+
+	if err := w.Add(cfgDir); err != nil {
+		return errors.Wrap(err, "add file watch")
+	}
+	if _, err := r.applyConfig(); err != nil {
+		return errors.Wrap(err, "initial apply")
+	}
+	if err := r.triggerReload(ctx); err != nil {
+		level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+	}
+
+	tick := time.NewTicker(3 * time.Minute)
+	defer tick.Stop()
+
+	for {
+		select {
+		case event := <-w.Events:
+			level.Debug(r.logger).Log("msg", "received watch event", "op", event.Op, "name", event.Name)
+
+			if event.Name != r.cfgFilename {
+				continue
+			}
+			changes, err := r.applyConfig()
+			if err != nil {
+				level.Error(r.logger).Log("msg", "apply failed", "err", err)
+			}
+			if !changes {
+				continue
+			}
+			if err := r.triggerReload(ctx); err != nil {
+				level.Error(r.logger).Log("msg", "triggering reload failed", "err", err)
+			}
+
+		case err := <-w.Errors:
+			level.Error(r.logger).Log("msg", "watch error", "err", err)
+
+		case <-tick.C:
+			changes, err := r.refreshRules()
```

Comment:
Essentially, file watches are a total mess and using them recursively on directories is a really bad idea. For rules we are probably totally fine with a 3 minute delay.